### PR TITLE
Remove message signatures when owner is removed

### PR DESCRIPTION
**Pull request by Uxio0**

### CODE REVIEW

File: safe_transaction_service/history/indexers/tx_processor.py [hunk 1]
```diff
@@ -235,9 +237,10 @@ def swap_owner(
         MultisigConfirmation.objects.remove_unused_confirmations(
             contract_address, safe_status.nonce, owner
         )
+        safe_message_models.SafeMessageConfirmation.objects.filter(owner=owner).delete()
```
Comment (moisses89) [inline safe_transaction_service/history/indexers/tx_processor.py:240]: If `SafeMessageConfirmation` have a owner `foreignkey` with delete on cascade shouldn't remove the `messageconfirmation` if the owner is removed?

Comment (fmrsabino) [reply safe_transaction_service/history/indexers/tx_processor.py:240]: `SafeMessageConfirmation` does not have a foreign key to the owner right? It has to the `SafeMessage`. If the `SafeMessage` is removed then the associated confirmations should be removed too.

File: safe_transaction_service/history/indexers/tx_processor.py [hunk 2]
```diff
@@ -235,9 +237,10 @@ def swap_owner(
...
 
     def store_new_safe_status(
-        self, safe_last_status: SafeStatus, internal_tx: InternalTx
+        self, safe_last_status: SafeLastStatus, internal_tx: InternalTx
```
Comment (fmrsabino) [inline safe_transaction_service/history/indexers/tx_processor.py:243]: What is the reason to change the type here?

File: safe_transaction_service/history/tests/test_tx_processor.py [hunk 3]
```diff
@@ -117,14 +123,23 @@ def test_tx_processor_with_factory(self):
         self.assertEqual(safe_status.threshold, threshold)
 
         threshold = 1
-        # Check deleting the owner did delete this pending confirmation
+        # Check deleting the owner did delete this pending confirmation for both signature and transaction
```
Comment (fmrsabino) [inline safe_transaction_service/history/tests/test_tx_processor.py:126]: nitpick: this test has around 274 lines. By adding functionality and editing the test it becomes more difficult to maintain the test.

My suggestion is to split the test into multiple ones where we test these conditions seperately. The setup might end up being the same but by adding more meaning to what is expected on each test it also makes it easier to debug.

File: safe_transaction_service/safe_messages/migrations/0002_alter_safemessageconfirmation_owner.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# Generated by Django 4.1.3 on 2022-11-16 14:35
+
+from django.db import migrations
+
+import gnosis.eth.django.models
+
+
+class Migration(migrations.Migration):
```
Comment (fmrsabino) [inline safe_transaction_service/safe_messages/migrations/0002_alter_safemessageconfirmation_owner.py:8]: What is the context for this migration. The model was not changed 🤔

File: safe_transaction_service/safe_messages/tests/factories.py [hunk 5]
```diff
@@ -33,4 +33,4 @@ class Params:
             encode_defunct(text=o.safe_message.message)
         )["signature"].hex()
     )
-    signature_type = 3
+    signature_type = 2
```
Comment (fmrsabino) [inline safe_transaction_service/safe_messages/tests/factories.py:36]: Why was this changed?

File: safe_transaction_service/safe_messages/serializers.py [hunk 6]
```diff
@@ -66,6 +66,11 @@ def validate(self, attrs):
             raise ValidationError(
                 f"1 owner signature was expected, {len(safe_signatures)} received"
             )
+        for safe_signature in safe_signatures:
+            if not safe_signature.is_valid(ethereum_client, safe_address):
+                raise ValidationError(
+                    f"Signature={safe_signature.signature.hex()} for owner={safe_signature.owner} is not valid"
+                )
```
Comment (fmrsabino) [inline safe_transaction_service/safe_messages/serializers.py:73]: What is the reason for this check? Was it missing?

Maybe `SafeSignature.parse_signature` should throw in these cases or?

File: safe_transaction_service/history/indexers/tx_processor.py (hunk 1 above)
Comment (Uxio0) [reply safe_transaction_service/history/indexers/tx_processor.py:240]: That's it, no foreign key to the owner

File: safe_transaction_service/history/indexers/tx_processor.py (hunk 2 above)
Comment (Uxio0) [reply safe_transaction_service/history/indexers/tx_processor.py:243]: It was a typo before and I detected it when checking the `tx_processor` for implementing this feature

File: safe_transaction_service/history/tests/test_tx_processor.py (hunk 3 above)
Comment (Uxio0) [reply safe_transaction_service/history/tests/test_tx_processor.py:126]: @fmrsabino This test checks a complete flow for indexing a Safe, and I cannot think of how to split as it depends on the status of the previous steps. This test is completely necesary and has been really useful for detecting indexing issues, but I agree that maybe we could have different tests for checking all the `tx_processor` options and branches

File: safe_transaction_service/safe_messages/migrations/0002_alter_safemessageconfirmation_owner.py (hunk 4 above)
Comment (Uxio0) [reply safe_transaction_service/safe_messages/migrations/0002_alter_safemessageconfirmation_owner.py:8]: I forgot to add the migration in the previous PR. Either way I will reset migrations before going to production when modifying the `description` -> `safeAppId`

File: safe_transaction_service/safe_messages/tests/factories.py (hunk 5 above)
Comment (Uxio0) [reply safe_transaction_service/safe_messages/tests/factories.py:36]: Signature is an EOA and not a ETH_SIGN

Comment (fmrsabino) [reply safe_transaction_service/safe_messages/tests/factories.py:36]: But this is a factory? Or did you mean to change the default?

File: safe_transaction_service/safe_messages/serializers.py (hunk 6 above)
Comment (Uxio0) [reply safe_transaction_service/safe_messages/serializers.py:73]: > What is the reason for this check? Was it missing?
Yes

> Maybe SafeSignature.parse_signature should throw in these cases or?
No, as for EIP1271 and ApprovedHash signatures a network request is required to check, and maybe is not desired. So first we parse the signatures (only in structure) and then you can validate them

Comment (fmrsabino) [reply safe_transaction_service/safe_messages/serializers.py:73]: > > Maybe SafeSignature.parse_signature should throw in these cases or?
> 
> No, as for EIP1271 and ApprovedHash signatures a network request is required to check, and maybe is not desired. So first we parse the signatures (only in structure) and then you can validate them

Ok in my opinion we should review these cases (if in the other cases they should throw a validation error or not). If the result of a "failed" parse can be ignored in some cases then I agree that we don't need to change it.

**Review by fmrsabino — APPROVED** [review]

File: safe_transaction_service/safe_messages/tests/test_views.py [hunk 7]
```diff
@@ -303,6 +324,26 @@ def get_eip712_payload():
             },
         )
 
+        # Test not valid signature
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/safe_messages/tests/test_views.py]: I refactored the 2 tests using `subTest`

File: safe_transaction_service/safe_messages/tests/factories.py (hunk 5 above)
Comment (Uxio0) [reply safe_transaction_service/safe_messages/tests/factories.py:36]: No, is just a factory 😉

File: safe_transaction_service/safe_messages/serializers.py (hunk 6 above)
Comment (Uxio0) [reply safe_transaction_service/safe_messages/serializers.py:73]: But the problem is that parse can be valid but signature not, and that depends for example is a hash is approved on the contract, and you don't want to check the signatures when you parse them as blockchain calls are not really fast (we are using it in a lot of places as it's a core function from `safe-eth-py`)
